File: slicing-android/app/src/main/java/com/crescentflare/slicinggame/components/navigationbars/SimpleBottomBarView.kt
package com.crescentflare.slicinggame.components.navigationbars

import android.annotation.TargetApi
import android.content.Context
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.util.AttributeSet
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import com.crescentflare.slicinggame.components.types.NavigationBarComponent
import com.crescentflare.slicinggame.components.utility.ViewletUtil
import com.crescentflare.unilayout.views.UniView


/**
 * Navigation bar: used to colorize the space under the bottom navigation bar controls (like home, back, etc.)
 */
open class SimpleBottomBarView : UniView, NavigationBarComponent {

    // --
    // Statics
    // --

    companion object {

        // --
        // Static: viewlet integration
        // --

        val viewlet: JsonInflatable = object : JsonInflatable {
            override fun create(context: Context): Any {
                return SimpleBottomBarView(context)
            }

            override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
                if (obj is SimpleBottomBarView) {
                    ViewletUtil.applyGenericViewAttributes(mapUtil, obj, attributes)
                    return true
                }
                return false
            }

            override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
                return obj::class == SimpleBottomBarView::class
            }
        }
    }


    // --
    // Initialization
    // --

    @JvmOverloads
    constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0)
            : super(context, attrs, defStyleAttr)

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int)
            : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        // No implementation
    }


    // --
    // NavigationBarComponent implementation
    // --

    override val averageColor: Int?
        get() = (background as? ColorDrawable)?.color

    override var statusBarHeight: Int = 0
    override var barContentHeight: Int = 0

}

File: slicing-ios/SlicingGame/Components/NavigationBars/SimpleBottomBarView.swift
//
//  SimpleBottomBarView.swift
//  Navigation bar: used to colorize the space under the bottom safe area
//

import UIKit
import UniLayout
import JsonInflator

class SimpleBottomBarView: UniView, NavigationBarComponent {

    // --
    // MARK: Viewlet integration
    // --
    
    class func viewlet() -> JsonInflatable {
        return ViewletClass()
    }

    private class ViewletClass: JsonInflatable {

        func create() -> Any {
            return SimpleBottomBarView()
        }

        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            if let bottomBar = object as? SimpleBottomBarView {
                ViewletUtil.applyGenericViewAttributes(convUtil: convUtil, view: bottomBar, attributes: attributes)
                return true
            }
            return false
        }

        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return type(of: object) == SimpleBottomBarView.self
        }

    }


    // --
    // MARK: Initialization
    // --
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }

    private func setup() {
        // No implementation
    }


    // --
    // MARK: NavigationBarComponent implementation
    // --
    
    var averageColor: UIColor? {
        get {
            return backgroundColor
        }
    }

    var statusBarHeight: CGFloat = 0
    var barContentHeight: CGFloat = 0

}
